### PR TITLE
Add JQ_COLORS env var for color config (fix #1252)

### PR DESCRIPTION
**Pull request by nicowilliams**
This commit adds an env var, `JQ_COLORS`, that allows the user to specify an alternative color scheme.

You can't see it here, but this produces something of a rainbow:

```
$ JQ_COLORS="1;30:0;31:0;32:0;33:0;34:1;35:1;36" ./jq -Ccn '[{"a":true,"b":false},123,null]'
[{"a":true,"b":false},123,null]
$ 
```

### CODE REVIEW

**Comment by nicowilliams** [comment]
This addresses #1252.

File: src/jv_print.c [hunk 1]
```diff
@@ -27,11 +27,46 @@
...
+    if ((e = strchr(c, ':')) == NULL)
+      e = c + strlen(c);
+    if ((size_t)(e - c) > sizeof(color_bufs[i]) - 1)
+      return 0;
+    color_bufs[i][0] = ESC[0];
+    color_bufs[i][1] = '[';
+    (void) strncpy(&color_bufs[i][2], c, e - c);
+    if (strspn(&color_bufs[i][2], "0123456789;") < strlen(&color_bufs[i][2]))
+      return 0;
+    color_bufs[i][2 + e - c] = 'm';
```
Comment (wtlangford) [inline src/jv_print.c]: Also, you're not guaranteed to write a `NUL` byte into `color_bufs[i][3 + e - c]`.  Let's explicitly do that so we can't run off the end later.

File: tests/shtest [hunk 2]
```diff
@@ -244,4 +244,41 @@ if [ "`$VALGRIND $Q $JQ -n '"xyz\n"|halt_error(1)' 2>&1`" != xyz ]; then
...
+$JQ -Ccn '[{"a":true,"b":false},123,null]' > $d/expect
+JQ_COLORS='garbage;30:*;31:,;3^:0;$%:0;34:1;35:1;36' \
+  $JQ -Ccn '[{"a":true,"b":false},123,null]' > $d/color
+cmp $d/color $d/expect
+JQ_COLORS='1234567890123456789;30:0;31:0;32:0;33:0;34:1;35:1;36' \
+  $JQ -Ccn '[{"a":true,"b":false},123,null]' > $d/color
+cmp $d/color $d/expect
+JQ_COLORS='1;31234567890123456789:0;31:0;32:0;33:0;34:1;35:1;36' \
+  $JQ -Ccn '[{"a":true,"b":false},123,null]' > $d/color
+cmp $d/color $d/expect
```
Comment (wtlangford) [inline tests/shtest]: Let's also add a test where we expect failure with a strings like `12345678901234:12345678901234:12345678901234:12345678901234:12345678901234:12345678901234:12345678901234` (for buffer overrun)
and `1234567890123:1234567890123:1234567890123:1234567890123:1234567890123:1234567890123:1234567890123` (for lack of trailing NUL)

File: src/jv_print.c [hunk 3]
```diff
@@ -27,11 +27,46 @@
 static const jv_kind color_kinds[] =
   {JV_KIND_NULL,   JV_KIND_FALSE, JV_KIND_TRUE, JV_KIND_NUMBER,
    JV_KIND_STRING, JV_KIND_ARRAY, JV_KIND_OBJECT};
-static const char* const colors[] =
+static char color_bufs[8][16];
```
Comment (nicowilliams) [inline src/jv_print.c]: Uh, no reason.  I'll change it.

File: src/jv_print.c [hunk 4]
```diff
@@ -27,11 +27,46 @@
...
+int
+jq_set_colors(const char *c)
+{
+  const char *e;
+  size_t i;
+
+  if (c == NULL)
+    return 1;
+  colors = def_colors;
+  memset(color_bufs, 0, sizeof(colors));
```
Comment (nicowilliams) [inline src/jv_print.c]: Oy, yes.

File: src/jv_print.c [hunk 5]
```diff
@@ -27,11 +27,46 @@
...
+  for (i = 0; i < sizeof(def_colors) / sizeof(def_colors[0]); i++)
+    color_bufps[i] = def_colors[i];
+  for (i = 0; i < sizeof(def_colors) / sizeof(def_colors[0]) && *c != '\0'; i++, c = e) {
```
Comment (nicowilliams) [inline src/jv_print.c]: Will fix.

File: src/jv_print.c [hunk 6]
```diff
@@ -27,11 +27,46 @@
...
+    if ((e = strchr(c, ':')) == NULL)
+      e = c + strlen(c);
+    if ((size_t)(e - c) > sizeof(color_bufs[i]) - 1)
```
Comment (nicowilliams) [inline src/jv_print.c]: Thanks!  (Though I attempted to test this.)

File: src/jv_print.c (hunk 1 above)
Comment (nicowilliams) [inline src/jv_print.c]: That's what the memset() is for.

File: tests/shtest (hunk 2 above)
Comment (nicowilliams) [inline tests/shtest]: Good idea.

**Comment by nicowilliams** [comment]
Fixes pushed.

File: src/jv_print.c [hunk 7]
```diff
@@ -27,11 +27,46 @@
...
+      return 0;
+    color_bufs[i][0] = ESC[0];
+    color_bufs[i][1] = '[';
+    (void) strncpy(&color_bufs[i][2], c, e - c);
```
Comment (nicowilliams) [inline src/jv_print.c]: It certainly doesn't overflow now (there's a test).

**Comment by nicowilliams** [comment]
Thanks!

**Pull request merged**